### ISSUE TITLE
about undefined properties in tsoa request body

### ISSUE THREAD

**Issue by simllll** [lukeautry/tsoa#525]
## Sorting

* **I'm submitting a ...**
  - [X] bug report
  - [ ] feature request
  - [X] support request

* I confirm that I
  - [X] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Issue / Question

I have a simple input post body e.g.: 
```ts
interface InputDTO {
	firstName?: string;
	lastName?: string;
	email?: string;
	phone?: string;
	birthday?: Date;
	gender?: Gender;
	nationality?: string;
}
 enum Gender {
	Male = 'male',
	Female = 'female'
}

     @Post('/test')
	async test(
		@Request() req,
		@Body() body: InputDTO
	): Promise<{
		userUpdate: { [key: string]: any };
	}> {
		console.log('body', body);
       }
```

and if I call it with "firstName" only, all other properties are stil there, but undefined.
```js
{
  firstName: 'safd',
  lastName: undefined,
  email: undefined,
  phone: undefined,
  birthday: undefined,
  gender: undefined,
  nationality: undefined
}
```

the question is, is this expected? Therefore I cannot distinguish between calling "firstName: undefined" and firstName not provided at all.
The check 
```js 
if ('firstName' in body) {
    console.log('firstName provided');
}
``` 
is always true.

## Possible Solution
Only set values if they are provided and set undefined as fallback if the property exists in the request but has no value.

## Context (Environment)
Version of the library: tsoa@2.5.8
Version of NodeJS: v12.13.0

**Comment by dgreene1** [comment]
> Therefore I cannot distinguish between calling "firstName: undefined" and firstName not provided at all.

@simllll can you describe the benefit of being able to distinguish between the two cases?

Because in both cases the consumer didn’t send a value. And since your swagger docs declare that it’s optional, then they don’t have to send it, right?

I’m asking this because while I appreciate the value of accuracy, this seems like a low priority change. However, if you can clarify the benefit then it could increase the priority of the ticket.

**Comment by simllll** [comment]
Basically we have a method here that sets values depending on that what is provided. E.g. only set a firstname if provided, set gender if defined or unset values if they are undefined.
For example the gender: It can actually have 4 states with this:
* it is provided and has a value of "male"
* it is provided and has a value of "female"
* It is provided, but undefined .. remove gender
* it is not provided - do not do anything with the gender property.

Right now this leads to the behaviour that all non provided properties get removed after each call. e.g. fist call sets firstname, second call sets lastname (but firstname is still defined, but undefined) and therefore removes firstname again.

This could be easily solved for strings, but for enums this is a bigger problem. I would have to change the whole application logic for unsetting the gender for example.

**Comment by dgreene1** [comment]
If I'm understanding you correctly, the end result for anything other than ”male” or ”female” is that you transform it to not set.

So if the end result is the same, I still don't see why it's important to know if the consumer set it or not.

Btw, if you'd like to submit a PR for this I think it's a reasonable ask.

But if you take on this work, please consider this special case.

```ts
export interface NotProvidedInterfaceExample {
  mustProvide: undefined, 
  optional: string | undefined
  canBeMissing?: string
}

const payload: NotProvidedInterfaceExample ={
  mustProvide: undefined, 
  optional: undefined
};

const response = await send(”api/undefinedPropertyTest”, payload)
const body = response.body;

expect(body.hasOwnProperty(”mustProvide”)).to.equal(true)
expect(body.hasOwnProperty(”optional”)).to.equal(true)
expect(body.hasOwnProperty(”canBeMissing”)).to.equal(false)
```

**Comment by simllll** [comment]
> If I'm understanding you correctly, the end result for anything other than ”male” or ”female” is that you transform it to not set.

no, not exaclty, my check for this is just this simple line:
```
if ('gender' in update) {
			// update gender
			userUpdate.gender = update.gender;
		}
```

'gender' in update checks if the gender is defined (can be male, female or undefined). If if is undefined, it will be removed.

Anyways, thanks for your input, I will look If I'm capable of solving this with a PR :)

**Comment by simllll** [comment]
I've added a PR #526, let me know what you think. Thanks a lot

**Comment by dgreene1** [comment]
@simllll since you said that you really needed this, I just released the fix as part of v2.5.9. Thank you for the fix. Thank you for being an active member of the tsoa community. We have other open PRs that would really benefit from your help if you're interested. Thanks again! :)

**Comment by simllll** [comment]
Thanks a lot @dgreene1 

Are you talking about a specific PR / bug / feature request?

**Comment by dgreene1** [comment]
> Are you talking about a specific PR / bug / feature request?

Nope, nothing specific. I was just expressing gratitude for the contribution and requesting more contributions if you're open to it. If you're interested, I've added a "help wanted" label for any PR that is clearly defined, and I've added an "Easy First PR" label for PRs that are more achievable. Thanks again! I hope to see you around on these boards. :)
